examples/jsm/renderers/webgpu/WebGPURenderer.js

```diff
@@ -18,9 +20,23 @@ class WebGPURenderer extends Renderer {
 
 	constructor( parameters = {} ) {
 
-		const backend = new WebGPUBackend( parameters );
-		//const backend = new Proxy( new WebGPUBackend( parameters ), debugHandler );
+		let BackendClass;
 
+		if ( WebGPU.isAvailable() ) {
+
+			BackendClass = WebGPUBackend;
+
+		} else {
+
+			BackendClass = WebGLBackend;
```

Comment:
I again suggest we should not do this.
WebGPURenderer running WebGL would be **extremely** confusing for the users.
We can make WebGPURenderer using WebGPUBackend, WebGLRenderer using WebGLBackend, and something like UniversalRenderer that uses WebGPUBackend and can fall through to WebGLBackend. Or we can just advise users to do `new Renderer( any backend they want to use )`.

Reply:
`webgpu_*` examples should also only showcase WebGPUBackend examples.

Reply:
> WebGPURenderer running WebGL would be extremely confusing for the users.

Do you mind elaborating?

> I again suggest we should not do this. WebGPURenderer running WebGL would be **extremely** confusing for the users.

Do you mind elaborating which a couple of examples?

Reply:
We're doing this to move people to WebGPU without them having to worry that is not yet supported in Firefox, Safari, Android, iOS and Linux. Eventually these platforms will support WebGPU and then the WebGPU backend will be used automatically without having to change anything.

Most of the developers using the library do not really know the difference between WebGL and WebGPU, neither should they. The people that know what they difference is will probably not have a hard time understanding that there's a temporary WebGL backend in place.

Having said that `WebGPURenderer` is still being worked on and it is not ready to replace `WebGLRenderer` just yet.

Reply:
> Do you mind elaborating with a couple of examples?

As a user, I expect that if I see a class named WebGPURenderer, it means it uses WebGPU -- i.e., I can write WGSL shaders for it, I can use bleeding-edge WebGPU features, I don't encounter WebGL bugs. If all out of sudden after many hours of debugging "Why this doesn't work?" I'll realize that all this time the name of the class was lying and actually it was WebGL -- I would be extremely confused.

> We're doing this to move people to WebGPU without them having to worry that is not yet supported in Firefox, Safari, Android, iOS and Linux.

This is a **too large thing** that people could "not worry about". They cannot expect perfect compatibility between WebGL and WebGPU! They **must** carefully plan if they are going to use WebGL, WebGPU, or both -- we shouldn't pack a combination between WebGL and WebGPU into a box called WebGPU. At most we can only create some UniversalRenderer and make it explicit that it could use both backends and people should think about and test it very carefully.

> Most of the developers using the library do not really know the difference between WebGL and WebGPU, neither should they.

Both parts of this sentence are wrong. Developers **should** (and do) know that Three.js' WebGLRenderer is based on a graphics API called WebGL and WebGPURenderer based on a graphics API called WebGPU, and they **should** know (and we should explicitly say it in docs and tutorials!) that these APIs **may not** be perfectly compatible.

> Having said that WebGPURenderer is still being worked on and it is not ready to replace WebGLRenderer just yet.

This is why I hope the decision to have WebGPURenderer with WebGLBackend could still be overturned and we could consider other alternatives (such as an UniversalRenderer).

Reply:
Sorry, but this has already been discussed and decision was made. Why do we now restart this topic again?

Reply:
The last time we were discussing this the discussion was stopped and it was suggested to discuss that in a new issue (because that one wasn't really related to this topic).

Reply:
> The last time we were discussing this the discussion was stopped

Can you please share a link to this discussion? AFAIK, there was a clear decision by @mrdoob do go for the current design (for good reasons).

Reply:
From what I understand the timeline would be this:

![image](https://github.com/mrdoob/three.js/assets/502810/0dce1fd3-bb02-4ab5-a4bf-1486b96deadc)

I think for specific purposes we still can use `new Renderer( new WebGPUBackend() )`, this would not affect the operation in any way.

Reply:
> Can you please share a link to this discussion?

I think it was https://github.com/mrdoob/three.js/pull/25750#discussion_r1155780519. I don't see there the message about moving the discussion to a new issue, maybe it was deleted.

> go for the current design (for good reasons).

What are the reasons _in favor_ of having WebGPURenderer using WebGLBackend (instead of, for example, having a class that does **exactly** the same, but is called, for example, UniversalRenderer)? I gave reasons against that in the old discussion and in this discussion again.

Reply:
> As a user, I expect that if I see a class named WebGPURenderer, it means it uses WebGPU -- i.e., I can write WGSL shaders for it, I can use bleeding-edge WebGPU features, I don't encounter WebGL bugs. If all out of sudden after many hours of debugging "Why this doesn't work?" I'll realize that all this time the name of the class was deceiving me and actually it was WebGL -- I would be extremely confused, to say the least.

That has a pretty easy solution.

When the renderer uses the WebGLBackend we should just console log  that the system doesn't support WebGPU and that the renderer is using WebGL2 as fallback and that some features may not be supported. 

Later on we can specify exactly which features are not supported when we know.

Reply:
> From what I understand the timeline would be this:
> 
> ![image](https://user-images.githubusercontent.com/502810/265753781-0719a5a2-5dd6-4d3d-ae10-4efa9617e1cc.png)

Haha!

That graph is missing the `CanvasRenderer` before `WebGLRenderer` and the `SVGRenderer` before that too.

Can't remember the name of the renderer class I picked when it was in ActionScript 😁

Reply:
About the timeline -- I think the important thing there is that "WebGLRenderer -- WebGL 2/WebGL" was OK, because WebGL 2 was backwards compatible with WebGL. WebGPU isn't.

> That has a pretty easy solution.
>
> When the renderer uses the WebGLBackend we should just console log that the system doesn't support WebGPU and that the renderer is using WebGL2 as fallback and that some features may not be supported.
>
> Later on we can specify exactly which features are not supported when we know.

Okay. But I'd expect the feature list (supported in WebGPU but not in WebGL, bugs not presented in WebGPU but that presented in WebGL) to be quite large. And therefore if a user uses such a feature they generally expect that only WebGPU is used (as they may expect from the title)...

But mostly I'm arguing only about the name. If we just rename (current) WebGPURenderer to UniversalRenderer, or WebGPUWithWebGLFallbackRenderer, or anything like that, I'll be fine.

Reply:
I do not think `WebGPURenderer` should fall back to `WebGLRenderer`. I think the app should handle that.

Reply:
`WebGPURenderer` with the new material system follows a different direction than the previous renderers. So insisting to keep the same policy as before ignores the new project's focus.

The node material in combination with TSL will decrease the number of occasions where devs want to write native shader code. In the end, it really doesn't matter on what backend the app is rendered. This will make it easier for the average users to develop good web experiences since they don't have to take care about fallbacks. The renderer and the material system will do this automatically. This supports the most important goal of the project: Make the 3D in the web easier to use.

Reply:
> WebGPURenderer with the new material system follows a different direction than the previous renderers.

But this is true not because it uses WebGPU or WebGL but because it is based on the Nodes system.

I'm not against an universal renderer (although I think most people wouldn't use it) but again, I think it shouldn't be named WebGPURenderer.

Reply:
I think the name is fine since the renderer primarily targets `WebGPU`.